### PR TITLE
Added JSON abilities

### PR DESCRIPTION
**Pull request by clementpeihengtan**

### CODE REVIEW

File: Cogs.Console/Cogs.Console.csproj [hunk 1]
```diff
@@ -19,4 +20,10 @@
     <ProjectReference Include="..\Cogs.Publishers\Cogs.Publishers.csproj" />
   </ItemGroup>
 
+  <ItemGroup>
+    <Reference Include="System.Web.Extensions">
+      <HintPath>..\..\..\..\..\..\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.0\System.Web.Extensions.dll</HintPath>
```
Comment (DanSmith) [inline Cogs.Console/Cogs.Console.csproj]: Is this reference assembly used anymore?

File: Cogs.Console/Properties/launchSettings.json [hunk 2]
```diff
@@ -0,0 +1,8 @@
+{
+  "profiles": {
+    "Cogs.Console": {
+      "commandName": "Project",
+      "commandLineArgs": "publish-json -o \"C:\\Users\\clement\\Documents\\GitHub\\cogs\\cogsburger\""
```
Comment (DanSmith) [inline Cogs.Console/Properties/launchSettings.json]: Lets ignore this file from the git repo

File: Cogs.Publishers/Cogs.Publishers.csproj [hunk 3]
```diff
@@ -9,7 +9,17 @@
...
+    <PackageReference Include="Newtonsoft.Json" Version="10.0.2" />
+  </ItemGroup>
+
+  <ItemGroup>
     <ProjectReference Include="..\Cogs.Model\Cogs.Model.csproj" />
   </ItemGroup>
 
+  <ItemGroup>
+    <Reference Include="System.Web.Extensions">
+      <HintPath>..\..\..\..\..\..\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5.2\System.Web.Extensions.dll</HintPath>
```
Comment (DanSmith) [inline Cogs.Publishers/Cogs.Publishers.csproj]: Is this reference assembly used anymore? I think json serialization switched to json.net

File: Cogs.Publishers/JsonPublisher.cs [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+                        //traverse parent list, find the properties of the parents
+                        foreach (var properti in item.ParentTypes)
+                        {
+                            if (properti.Properties != null)
+                            {
+                                //traverse the properties and get all the information regarding variable. 
+                                foreach (var inner_prop in properti.Properties)
+                                {
+                                    var parentprop = new JsonSchemaProp();
+                                    parentprop.name = inner_prop.Name;
```
Comment (DanSmith) [inline Cogs.Publishers/JsonPublisher.cs]: Lets refactor getting the properties from the parents and the items, so the property creation isn't repeated.

We could make a local list of Properties, add the Properties from the parents and the item in the correct order, and then convert them all at once to JsonSchemaProps

File: Cogs.Publishers/JsonScehmaPropConverter.cs [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        }
+
+        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+        {
+            throw new NotImplementedException();
+        }
+
+        public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
+        {
+            if (value.GetType() == typeof(List<JsonSchemaProp>))
```
Comment (DanSmith) [inline Cogs.Publishers/JsonScehmaPropConverter.cs]: In c# 7 there is a new feature that makes this type of type testing easier called pattern matching
https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-7#pattern-matching

We can change the if statement to 
if(value is List<JsonSchemaProp>prop_list) { .... } to test the type and declare a typed variable at the same time

File: Cogs.Publishers/JsonSchema.cs [hunk 6]
```diff
@@ -0,0 +1,16 @@
+﻿using Newtonsoft.Json;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Cogs.Publishers
+{
+    class JsonSchema
+    {
+        public string title;
```
Comment (DanSmith) [inline Cogs.Publishers/JsonSchema.cs]: Lets convert these to get and set properties, and use camel casing. Such as

public string Title { get; set; }

We can do this in both of these classes

File: Cogs.Publishers/JsonSchemaProp.cs [hunk 7]
```diff
@@ -0,0 +1,11 @@
+﻿namespace Cogs.Publishers
+{
+    public class JsonSchemaProp
+    {
+        public string name; //TODO: implement own serializer
+        public string type;
```
Comment (DanSmith) [inline Cogs.Publishers/JsonSchemaProp.cs]: Lets convert these to get and set properties, and use camel casing. Such as

public string Type { get; set; }

We can do this in both of these classes

File: Cogs.Publishers/JsonSchema.cs [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+        public string type;
+        [JsonConverter(typeof(JsonScehmaPropConverter))]
+        public List<JsonSchemaProp> properties { get; set; }
```
Comment (jeremyiverson) [inline Cogs.Publishers/JsonSchema.cs]:
```csharp
public List<JsonSchemaProp> Properties { get; } = new List<JsonSchemaProp>();
```

**Pull request merged**